### ISSUE TITLE
Long time reading a file with scv.read

### ISSUE THREAD

**Issue by SamueleSoraggi** [theislab/scvelo#25]
Hi!
I am trying to read a .loom file, whose size is approx 900MB. I have installed scvelo v0.1.13 and I have used the command

`scvelo.read(myFile, sparse=True, cache=True)`

The command seems to take a long time (it has been running for almost one hour), but when I look at the job running on my cluster, I can see no more data gets loaded into memory.
I was wondering id such a long time has been happening already to someone using scvelo, or if it can be caused by dependencies.

Thanks for help, you guys are doing great job 👍 
Samuele

**Comment by VolkerBergen** [comment]
Hi Samuele,

That should not take more than a few minutes when loading for the first time (after writing the cache it becomes even faster). We are working with data which is much larger than this. 

Could you check the dependencies with `scvelo.logging.print_versions()`? In particular, your versions of AnnData and loompy would be helpful to know. 

There might also be something wrong with the already written cache file. You could further try to set `cache=False`.

**Comment by SamueleSoraggi** [comment]
`import loompy
print(loompy.loom_spec_version)`

returns 2.0.1

`import anndata
anndata.__version__`

returns 0.6.11

I will try to set `cache=False` in the meanshile, to see if something changes :)

**Comment by VolkerBergen** [comment]
That looks all fine. Can you check whether you can read a predefined dataset with 
`adata = scvelo.datasets.dentategyrus()`

**Comment by SamueleSoraggi** [comment]
It works in something less than a minute and I get

>>> adata = scv.datasets.dentategyrus()
... 100%... storing 'Chromosome' as categorical
... storing 'Strand' as categorical
... storing 'clusters' as categorical

**Comment by VolkerBergen** [comment]
Alright, that basically means that there is nothing wrong with the actual `scv.read` function nor dependencies, but it has most likely something to do with your data.

Could you now check whether the loompy connection works, i.e. `ds = loompy.connect(filename)`.

**Comment by SamueleSoraggi** [comment]
It seems doing fine. If I check the variable ds i get

'<loompy.loompy.LoomConnection object at 0x7f3eabc358d0>'

I am wondering if something is wrong with how I ran the velocyto command line tool on my data.

**Comment by VolkerBergen** [comment]
Now check the following commands (maybe line by line to see which one gives an error):
```
from collections import OrderedDict
from loompy import connect

lc = connect(filename)
X = lc.layers['spliced'].sparse().T.tocsr()

layers = OrderedDict()
for key in lc.layers.keys():
        if key != '': layers[key] = lc.layers[key].sparse().T.tocsr()

obs=dict(lc.col_attrs)
if 'CellID' in obs.keys(): obs['obs_names'] = obs.pop('CellID')

var=dict(lc.row_attrs)
if 'Gene' in var.keys(): var['var_names'] = var.pop('Gene')

adata = scv.AnnData(X, obs=obs, var=var, layers=layers, dtype=dtype)

lc.close()
```

**Comment by VolkerBergen** [comment]
If that works fine, also check:
```
from loompy import connect
with connect(filename, 'r') as lc:
        X = lc.layers['spliced'].sparse().T.tocsr()
```

**Comment by VolkerBergen** [comment]
Interesting.. either something wrong with the spliced matrix or it is not capable of making it sparse. Can you 
- print `lc.layers.keys() ` to check whether the spliced count matrix is available,
- print `lc.layers['spliced'].T` to see what it contains,
- if that works, print `lc.layers['spliced'].sparse().T.tocsr()` to check whether sparse() is the issue.

**Comment by SamueleSoraggi** [comment]
```python
>>> lc.layers.keys()
 ['', 'ambiguous', 'spliced', 'unspliced']
>>> lc.layers['spliced'].T
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'LoomLayer' object has no attribute 'T'
```

**Comment by VolkerBergen** [comment]
Oh right, that would be `lc.layers['spliced'][:].T` instead.

**Comment by SamueleSoraggi** [comment]
Ouch 🗡 
```python
>>> lc.layers['spliced'][:].T
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object is not subscriptable
```

By the way, after that for cycle (it just ended) I was able to run the subsequent commands very quickly.

**Comment by VolkerBergen** [comment]
Ahaa, so it seems to be `None` for whatever reason..
One more command to try: `lc.layers['spliced'][()].T`

**Comment by SamueleSoraggi** [comment]
Ah no wait, when I did it I had already run `lc.close()` !!!!! Ok I will try tomorrow again with less tired brain, so I do not risk wasting your time now doing stupid errors :)

**Comment by VolkerBergen** [comment]
No worries, then we'll solved that tmr! :)

**Comment by SamueleSoraggi** [comment]
Thanks for helping :) I owe you a beer

**Comment by VolkerBergen** [comment]
:D I'm totally down for it, when you're around!

**Comment by SamueleSoraggi** [comment]
Hi, good morning. I tried again all the commands and checked approx the time taken. Surprisingly, it works faster - really weird. This is what I get, it seems doing well also with sparse matrices.
This part worked in around 18 minutes, so faster than yesterday:

```python
>>> from collections import OrderedDict
>>> from loompy import connect
>>> import scvelo as scv
>>> filename = 'donor1.merged.loom'
>>> lc = connect(filename)
>>> X = lc.layers['spliced'].sparse().T.tocsr()  #5 minutes
>>> layers = OrderedDict()   
>>> for key in lc.layers.keys(): #13 minutes
>>>         if key != '': layers[key] = lc.layers[key].sparse().T.tocsr()
>>> obs=dict(lc.col_attrs)
>>> if 'CellID' in obs.keys(): obs['obs_names'] = obs.pop('CellID')
>>> var=dict(lc.row_attrs)
>>> if 'Gene' in var.keys(): var['var_names'] = var.pop('Gene')
>>> adata = scv.AnnData(X, obs=obs, var=var, layers=layers)
Variable names are not unique. To make them unique, call `.var_names_make_unique`.
>>> lc.close()
```

The next part took around 8 minutes (again, yesterday it took ages):
```python
>>> from loompy import connect
>>> with connect(filename, 'r') as lc:
>>>        X = lc.layers['spliced'].sparse().T.tocsr() # 8minutes
```

Also here does well:
```python
>>> lc = connect(filename)
>>> lc.layers.keys()
['', 'ambiguous', 'spliced', 'unspliced']
>>> lc.layers['spliced']
<loompy.loom_layer.LoomLayer at 0x2b2b816dd978>
>>> lc.layers['spliced'].sparse().T.tocsr() #7 minutes
<144184x33694 sparse matrix of type '<class 'numpy.uint16'>'
	with 18950157 stored elements in Compressed Sparse Row format>
```

I ran the same command as yesterday to load my dataset with `scvelo`, and works like a charm in literally 1 or 2 minutes
```python
adata2 = scv.read("donor1.merged.loom",sparse=True,cache=True)
```

So I do not really know how it happened that it was so slow. I could think of a very loaded node in our cluster, but I logged out and accessed again to get a different node assigned, and everything was still working slow. I will have to work with other 4 datasets on scvelo, so maybe we can have this issue opened only this couple of days - and if everything is fine we can close it?
Sorry for taking your time - but I really tried many times yesterday before writing the issue - I hope however it will be a good reference for other users with problems, and that I can offer you that beer if I happen to be around :)

Cheers,
Samuele

**Comment by VolkerBergen** [comment]
Your dataset is indeed relatively big. Thus 18 min for first run and 1-2 min from cache might make sense depending on the cluster. To further speedup you can read from cleaned data, as in 

```
scv.utils.cleanup(adata, clean='layers')
adata.write(filename)
```

which will throw away the matrices 'matrix' and 'ambiguous' that are redundant for velocity analysis and take quite some memory when having more than 100k cells.

Further when you already have your embeddings, annotations etc. done, you won't need to rerun this with scvelo, but simply merge them via `scv.utils.merge(adata, adata_loom)`. 

Just mentioning as these are not explained in the docs. 

Keep me up to date.

**Comment by SamueleSoraggi** [comment]
Hi,

I tried with all my datasets and the slowest was the one I worked with yesterday. It is also the largest. It took some 20 minutes for it, and between 3 and 15 minutes for the other 4 datasets, that are quite smaller. All of them load superquickly thanks to the cache the second time.

So it is nice just to know one can let them being opened in the background while doing something else, and have them cached for the next times they are needed.

Thanks for the help and also for the merge utility hint, I was gonna ask you right now about it! I learned some new stuff from this issue. I guess it can be closed :) Thanks a lot :) :beers:
